docs: init (#1025)
revanced/revanced-patches: docs/4_structure_and_conventions.md

# 📜 Patch file structure and conventions

> ReVanced follows couple conventions when creating patches which can be found in [ReVanced patches](https://github.com/revanced/revanced-patches).

Reviewer: ```suggestion
ReVanced follows a couple of conventions when creating patches which can be found in the [revanced-patches](https://github.com/revanced/revanced-patches) repository.
```

## 📁 File structure

Each patch is structured the following way:

```text

> 📦your.patches.app.categoriy.patch

Reviewer: ```suggestion
📦your.patches.app.category.patch
```

 ┣ 📂annotations
 ┣ ┗ ⚙️SomePatchCompatibility.kt
 ┣ 📂fingerprints
 ┣ ┣ 🔍SomeFingerprintA.kt
 ┣ ┗ 🔍SomeFingerprint.kt
 ┣ 📂patch
 ┗ ┗ 🧩SomePatch.kt
```

### 🆗 Example

> As an example the structure of [`VideoAdsPatch`](https://github.com/revanced/revanced-patches/tree/main/src/main/kotlin/app/revanced/patches/youtube/ad/video) can be used as a reference:

Reviewer: ```suggestion
As an example the structure of [`VideoAdsPatch`](https://github.com/revanced/revanced-patches/tree/2d10caffad3619791a0c3a670002a47051d4731e/src/main/kotlin/app/revanced/patches/youtube/ad/video) can be used as a reference:
```

```text
📦app.revanced.patches.youtube.ad.video
 ┣ 📂annotations
 ┣ ┗ ⚙️VideoAdsCompatibility.kt
 ┣ 📂fingerprints
 ┣ ┗ 🔍LoadVideoAdsFingerprint.kt
 ┣ 📂patch
 ┗ ┗ 🧩VideoAdsPatch.kt
```

## 📙 Conventions

> **Note**: More ⭐ equals more importance

- ⭐⭐ **`@Patch` should be named by what they accomplish**. Example: _To patch ads on videos, the patch should be called_VideoAdsPatch._

- ⭐⭐ **`@Description` should be written in third person and end with punctuation**. Example: _Removes ads in the video player._

> - ⭐ **Resource and bytecode patches should be properly seperated**. That means, bytecode patches handle patching bytecode, while resource patches handle resources. As an example, [`SponsorBlockPatch`](https://github.com/revanced/revanced-patches/tree/main/src/main/kotlin/app/revanced/patches/youtube/layout/sponsorblock) can be used.

Reviewer: ```suggestion
- ⭐ **Resource and bytecode patches should be properly separated**. That means, bytecode patches handle patching bytecode, while resource patches handle resources. As an example, [`SponsorBlockPatch`](https://github.com/revanced/revanced-patches/tree/2d10caffad3619791a0c3a670002a47051d4731e/src/main/kotlin/app/revanced/patches/youtube/layout/sponsorblock) can be used.
```

> - ⭐⭐⭐ **Allocate as little code as possible in patches**. This reduces the risk of failing patches. In the example of [`SponsorBlockPatch`](https://github.com/revanced/revanced-patches/tree/main/src/main/kotlin/app/revanced/patches/youtube/layout/sponsorblock), most of the code logic is written in the [revanced-integrations](https://github.com/revanced/revanced-integrations). The patches now only insert references to public methods from the integrations which are merged into the app which is far better than writing huge patches.

Reviewer: ```suggestion
- ⭐⭐⭐ **Allocate as little code as possible in patches**. This reduces the risk of failing patches. In the example of [`SponsorBlockPatch`](https://github.com/revanced/revanced-patches/tree/2d10caffad3619791a0c3a670002a47051d4731e/src/main/kotlin/app/revanced/patches/youtube/layout/sponsorblock), most of the code logic is written in the [revanced-integrations](https://github.com/revanced/revanced-integrations). The patches now only insert references to public methods from the integrations which are merged into the app which is far better than writing huge bytecode patches.
```
